lib/spoom/deadcode/indexer.rb

```diff
@@ -0,0 +1,394 @@
+# typed: strict
+# frozen_string_literal: true
+
+module Spoom
+  module Deadcode
+    class Indexer < SyntaxTree::Visitor
+      extend T::Sig
+
+      sig { returns(String) }
+      attr_reader :path, :file_name
+
+      sig { returns(Index) }
+      attr_reader :index
+
+      sig { params(path: String, source: String, index: Index).void }
+      def initialize(path, source, index)
+        super()
+
+        @path = path
+        @file_name = T.let(File.basename(path), String)
+        @source = source
+        @index = index
+        @previous_node = T.let(nil, T.nilable(SyntaxTree::Node))
+        @names_nesting = T.let([], T::Array[String])
+        @nodes_nesting = T.let([], T::Array[SyntaxTree::Node])
+        @in_const_field = T.let(false, T::Boolean)
+        @in_opassign = T.let(false, T::Boolean)
+        @in_symbol_literal = T.let(false, T::Boolean)
+      end
+
+      # Visit
+
+      sig { override.params(node: T.nilable(SyntaxTree::Node)).void }
+      def visit(node)
+        return unless node
+
+        @nodes_nesting << node
+        super
+        @nodes_nesting.pop
+        @previous_node = node
+      end
+
+      sig { override.params(node: SyntaxTree::AliasNode).void }
+      def visit_alias(node)
+        reference_method(node_string(node.right), node)
+      end
+
+      sig { override.params(node: SyntaxTree::ARef).void }
+      def visit_aref(node)
+        super
+
+        reference_method("[]", node)
+      end
+
+      sig { override.params(node: SyntaxTree::ARefField).void }
+      def visit_aref_field(node)
+        super
+
+        reference_method("[]=", node)
+      end
+
+      sig { override.params(node: SyntaxTree::ArgBlock).void }
+      def visit_arg_block(node)
+        value = node.value
+
+        case value
+        when SyntaxTree::SymbolLiteral
+          # If the block call is something like `x.select(&:foo)`, we need to reference the `foo` method
+          reference_method(symbol_string(value), node)
+        when SyntaxTree::VCall
+          # If the block call is something like `x.select { ... }`, we need to visit the block
+          visit(value)
```

Comment:
Technically we don't really need to call `super` since we already have the `node.value` figured out and this is the only node in an `ArgBlock`. We can save a few instructions. But indeed, we can call `super` for ease of understanding 👍